scipy/stats/_stats_py.py

```diff
@@ -9865,6 +9877,98 @@ def _cdf_distance(p, u_values, v_values, u_weights=None, v_weights=None):
+    elif wy is None:
+        x, y, wx = _broadcast_arrays((x, y, wx), axis=axis, xp=xp)
+        x, y, wx = (xp.moveaxis(x, axis, -1), xp.moveaxis(y, axis, -1),
+                    xp.moveaxis(wx, axis, -1))
+    else:
+        x, y, wx, wy = _broadcast_arrays((x, y, wx, wy), axis=axis, xp=xp)
+        x, y, wx, wy = (xp.moveaxis(x, axis, -1), xp.moveaxis(y, axis, -1),
+                        xp.moveaxis(wx, axis, -1), xp.moveaxis(wy, axis, -1))
+
+    # technically, if only one of wx were infinite along `axis`, we could
+    # treat it as if that weight were unity and all the rest were zero.
+    wx = wx if wx is None else xp.where((wx < 0) | xp.isinf(wx), xp.nan, wx)
+    wy = wy if wy is None else xp.where((wy < 0) | xp.isinf(wy), xp.nan, wy)
+
+    return x, y, wx, wy, axis, xp
+
+
+def _cdf_distance(p, x, y, wx=None, wy=None, axis=0):
+    # vectorized version of _cdf_distance_ref.
+    x, y, wx, wy, axis, xp = _cdf_distance_iv(x, y, wx, wy, axis)
+
+    # special case: either array is empty
+    if xp_size(x) == 0 or xp_size(y) == 0:
+        return xp.full(x.shape[:-1], xp.nan, dtype=x.dtype)
+
+    ix = xp.argsort(x, axis=-1)
+    iy = xp.argsort(y, axis=-1)
+    x = xp.take_along_axis(x, ix, axis=-1)
+    y = xp.take_along_axis(y, iy, axis=-1)
+    wx = wx if wx is None else xp.take_along_axis(wx, ix, axis=-1)
+    wy = wy if wy is None else xp.take_along_axis(wy, iy, axis=-1)
+    xy = xp.concat((x, y), axis=-1)
+
+    nx = x.shape[-1]
+    ny = y.shape[-1]
+    n = xy.shape[-1]
+    ranks = _rankdata(xy, method='min')
+    rank_counts_x = xp.diff(ranks[..., :nx], prepend=1, append=n + 1, axis=-1)
+    rank_counts_y = xp.diff(ranks[..., -ny:], prepend=1, append=n + 1, axis=-1)
+
+    if wx is None:
+        cdf_vals_x = xp.linspace(0, 1, nx + 1, dtype=xy.dtype)
+    else:
+        cdf_vals_x = xp.cumulative_sum(wx, axis=-1, include_initial=True)
+        cdf_vals_x /= cdf_vals_x[..., -1:]
+
+    if wy is None:
+        cdf_vals_y = xp.linspace(0, 1, ny + 1, dtype=xy.dtype)
+    else:
+        cdf_vals_y = xp.cumulative_sum(wy, axis=-1, include_initial=True)
+        cdf_vals_y /= cdf_vals_y[..., -1:]
+
+    cdf_vals_x, rank_counts_x = xp.broadcast_arrays(cdf_vals_x, rank_counts_x)
+    cdf_vals_y, rank_counts_y = xp.broadcast_arrays(cdf_vals_y, rank_counts_y)
+    cdf_x = xp.repeat(xp_ravel(cdf_vals_x), xp_ravel(rank_counts_x), axis=-1)
+    cdf_y = xp.repeat(xp_ravel(cdf_vals_y), xp_ravel(rank_counts_y), axis=-1)
+    shape = xy.shape[:-1] + (-1,)
+    cdf_x = xp.reshape(cdf_x, shape)[..., :-1]
+    cdf_y = xp.reshape(cdf_y, shape)[..., :-1]
+
+    deltas = xp.diff(xp.sort(xy, axis=-1), axis=-1)
+    if p == 1:
+        res = xp.vecdot(xp.abs(cdf_x - cdf_y), deltas, axis=-1)
+    elif p == 2:
+        res = xp.sqrt(xp.vecdot(xp.square(cdf_x - cdf_y), deltas))
+    elif p % 2:
+        res = xp.vecdot(xp.abs(cdf_x - cdf_y)**p, deltas)**(1 / p)
+    else:
+        res = xp.vecdot((cdf_x - cdf_y)**p, deltas)**(1 / p)
+    return res
```

Comment:
I'd be happy to eliminate all the branches. Just folllowing the original implementation, but I don't think this is the time-consuming part...